File: tasks/python3/processTemporalLayer.py
from datetime import datetime
import isodate
import re
import traceback

def to_list(val):
    return [val] if not hasattr(val, 'reverse') else val

# Add duration to end date using
# ISO 8601 duration keys
def determine_end_date(key, date):
    return date + isodate.parse_duration(key)

# This method takes a layer and a temporal
# value and tranlates it to start and end dates
def process_temporal(wv_layer, value):
    dateFormat = "%Y-%m-%d"
    timeFormat = "%H:%M:%S"
    dateTimeFormat = "%Y-%m-%d %H:%M:%S"
    try:
        ranges = to_list(value)
        if "T" in ranges[0]:
            wv_layer["period"] = "subdaily"

        else:
            if ranges[0].endswith("Y"):
                wv_layer["period"] = "yearly"
            elif ranges[0].endswith("M"):
                wv_layer["period"] = "monthly"
            else:
                wv_layer["period"] = "daily"
        start_date = datetime.max
        end_date = datetime.min
        date_range_start, date_range_end, range_interval = [], [], []

        for range in ranges:
            start, end, interval = range.split('/')
            if wv_layer["period"] == "daily" \
            or wv_layer["period"] == "monthly" \
            or wv_layer["period"] == "yearly":
                start_date = min(start_date, datetime.strptime(start, dateFormat))
                end_date = max(end_date, datetime.strptime(end, dateFormat))
                if start_date:
                    startDateParse = datetime.strptime(start, dateFormat)
                    date_range_start.append(startDateParse.strftime(dateFormat) + "T" + startDateParse.strftime(timeFormat) + "Z")
                if end_date:
                    endDateParse = datetime.strptime(end, dateFormat)
                    date_range_end.append(endDateParse.strftime(dateFormat) + "T" + endDateParse.strftime(timeFormat) + "Z")
                if interval != "P1D":
                    end_date = determine_end_date(interval, end_date)
                range_interval.append(re.search(r'\d+', interval).group())

            # Subdaily Layers
            else:
                startTime = start.replace('T', ' ').replace('Z', '')
                endTime = end.replace('T', ' ').replace('Z', '')
                start_date = min(start_date, datetime.strptime(startTime, dateTimeFormat))
                end_date = max(end_date, datetime.strptime(endTime, dateTimeFormat))

                if start_date:
                    startTimeParse = datetime.strptime(startTime, dateTimeFormat)
                    date_range_start.append(startTimeParse.strftime(dateFormat) + "T" + startTimeParse.strftime(timeFormat) + "Z")
                if end_date:
                    endTimeParse = datetime.strptime(endTime, dateTimeFormat)
                    date_range_end.append(endTimeParse.strftime(dateFormat) + "T" + endTimeParse.strftime(timeFormat) + "Z")

                range_interval.append(re.search(r'\d+', interval).group())

            wv_layer["startDate"] = start_date.strftime(dateFormat) + "T" + start_date.strftime(timeFormat) + "Z"
            if end_date != datetime.min:
                wv_layer["endDate"] = end_date.strftime(dateFormat) + "T" + end_date.strftime(timeFormat) + "Z"
            if date_range_start and date_range_end:
                wv_layer["dateRanges"] = [{"startDate": s, "endDate": e, "dateInterval": i} for s, e, i in zip(date_range_start, date_range_end, range_interval)]
    except ValueError:
        raise Exception("Invalid time: {0}".format(range))
    except Exception as e:
        print(traceback.format_exc())
        raise Exception("Error processing temporal layer: {0}".format(e))
    return wv_layer

File: tasks/build-options/processTemporalLayer.js
const fs = require('fs')
const moment = require('moment')
const xml2js = require('xml2js')

const projDict = {
  'GIBS:geographic': 'epsg4326',
  'GIBS:arctic': 'epsg3413',
  'GIBS:antarctic': 'epsg3031',
  'GITC:geographic': 'epsg4326',
  'GITC:arctic': 'epsg3413',
  'GITC:antarctic': 'epsg3031',
  'GITC:webmercator': 'epsg3857'
}

function toList (val) {
  return val instanceof Array ? val : [val]
}

function rangesHelper (wvLayer, ranges) {
  const dateFormat = 'YYYY-MM-DD'
  const dateTimeFormat = 'YYYY-MM-DD HH:mm:ss'
  let startDate = moment.min()
  let endDate = moment.max()
  const dateRangeStart = []
  const dateRangeEnd = []
  const rangeInterval = []
  if (ranges && ranges[0] && ranges[0]._text && ranges[0]._text.includes('T')) {
    wvLayer.period = 'subdaily'
  } else {
    if (ranges && ranges[0] && ranges[0]._text && ranges[0]._text.endsWith('Y')) {
      wvLayer.period = 'yearly'
    } else if (ranges && ranges[0] && ranges[0]._text && ranges[0]._text.endsWith('M')) {
      wvLayer.period = 'monthly'
    } else {
      wvLayer.period = 'daily'
    }
  }
  for (const range of ranges) {
    if (!range._text || !range._text.length) continue
    const [start, end, interval] = range._text.split('/')
    if (
      wvLayer.period === 'daily' ||
      wvLayer.period === 'monthly' ||
      wvLayer.period === 'yearly'
    ) {
      startDate = moment.min(startDate, moment(start, dateFormat))
      endDate = moment.max(endDate, moment(end, dateFormat))
      if (start) {
        startDate = moment(start, dateFormat).format('YYYY-MM-DDTHH:mm:ss[Z]')
        dateRangeStart.push(startDate)
      }
      if (end) {
        endDate = moment(end, dateFormat).format('YYYY-MM-DDTHH:mm:ss[Z]')
      }
      if (interval !== 'P1D') {
        endDate = moment.utc(endDate).add(moment.duration(interval)).format('YYYY-MM-DDTHH:mm:ss[Z]')
        // For monthly products subtract 1 day
        if (wvLayer.period === 'monthly') {
          endDate = moment.utc(endDate).subtract(1, 'day').format('YYYY-MM-DDTHH:mm:ss[Z]')
        }
      }
      const regex = /\d+/g
      const match = regex.exec(interval)
      rangeInterval.push(match)
      if (endDate.endsWith('T00:00:00Z')) {
        endDate = endDate.replace('T00:00:00Z', 'T23:59:59Z')
      }
      dateRangeEnd.push(endDate)
    } else {
      // Subdaily Layers
      startDate = moment(start, dateTimeFormat).format('YYYY-MM-DDTHH:mm:ss[Z]')
      endDate = moment(end, dateTimeFormat).format('YYYY-MM-DDTHH:mm:ss[Z]')

      if (start) {
        dateRangeStart.push(startDate)
      }
      if (end) {
        dateRangeEnd.push(endDate)
      }

      rangeInterval.push(interval.match(/\d+/)[0])
    }

    wvLayer.startDate = dateRangeStart[0]
    wvLayer.endDate = dateRangeEnd[dateRangeEnd.length - 1]
  }
  return [dateRangeStart, dateRangeEnd, rangeInterval]
}

async function createDateRanges (url, wvLayer, cacheMode) {
  const fetchOpts = cacheMode === 'no-store' ? { cache: cacheMode } : undefined
  let invalidRanges = false
  let ranges = []
  const dateRangeStart = []
  const dateRangeEnd = []
  const rangeInterval = []
  let nextPage
  const parser = new xml2js.Parser()
  try {
    const describeDomainsResponse = await fetch(url, fetchOpts)
    if (describeDomainsResponse?.ok) {
      const describeDomainsText = await describeDomainsResponse?.text?.() || ''
      const describeDomainsJson = await parser.parseStringPromise(describeDomainsText)
      nextPage = describeDomainsJson?.Domains?.DimensionDomain?.[0]?.NextPage?.[0]
      const domain = describeDomainsJson?.Domains?.DimensionDomain?.[0]?.Domain?.[0] || ''
      const domains = domain.split(',')
      if (domain && domains?.length) {
        const formattedDomains = domains.map((d) => {
          return {
            _text: d
          }
        })
        ranges = toList(formattedDomains)
      } else {
        invalidRanges = true
      }
    }
  } catch (error) {
    invalidRanges = true
    console.error(`Error fetching ${url}: ${error}`)
  } finally {
    if (!invalidRanges) {
      const [dateRangeStartYear, dateRangeEndYear, rangeIntervalYear] = rangesHelper(wvLayer, ranges)
      dateRangeStart.push(...dateRangeStartYear)
      dateRangeEnd.push(...dateRangeEndYear)
      rangeInterval.push(...rangeIntervalYear)
    }
    // Fetch next page recursively if one exists
    if (nextPage) {
      const [dateRangeStartOut, dateRangeEndOut, rangeIntervalOut] = await createDateRanges(nextPage, wvLayer, cacheMode)
      dateRangeStart.push(...dateRangeStartOut)
      dateRangeEnd.push(...dateRangeEndOut)
      rangeInterval.push(...rangeIntervalOut)
    }
  }
  return [dateRangeStart, dateRangeEnd, rangeInterval]
}

async function processTemporalLayer (wvLayer, value, source = 'GIBS:geographic', cacheMode) {
  try {
    const describeDomainsAllUrl = `https://gibs.earthdata.nasa.gov/wmts/${projDict[source]}/best/1.0.0/${wvLayer.id}/default/250m/all/all.xml`
    let [dateRangeStart, dateRangeEnd, rangeInterval] = await createDateRanges(describeDomainsAllUrl, wvLayer, cacheMode)
    // Fall back to original ranges if the 'all' endpoint fetch failed
    if (!dateRangeStart) {
      const rangesBackup = toList(value)
      const backupRangesObj = rangesHelper(wvLayer, rangesBackup)
      dateRangeStart = backupRangesObj[0]
      dateRangeEnd = backupRangesObj[1]
      rangeInterval = backupRangesObj[2]
    }
    if (dateRangeStart.length && dateRangeEnd.length) {
      const dateRangesObj = dateRangeStart.map((s, i) => ({
        startDate: s,
        endDate: dateRangeEnd[i],
        dateInterval: rangeInterval[i]
      }))
      if (wvLayer.period === 'subdaily') {
        await fs.writeFileSync(`build/options/config/dateRanges/${wvLayer.id}.json`, JSON.stringify(dateRangesObj, null, 2))
      } else {
        wvLayer.dateRanges = dateRangesObj
      }
    }
  } catch (e) {
    throw new Error(`Error processing temporal layer ${wvLayer.id}: ${e}`)
  }
  return wvLayer
}

module.exports = {
  processTemporalLayer
}
